carapax-i18n/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+//! An i18n handler for carapax
+
+#![warn(missing_docs)]
+
+#[cfg(feature = "gettext")]
+mod gettext;
+#[cfg(feature = "gettext")]
+pub use gettext::*;
+
+#[cfg(feature = "fluent")]
+#[macro_use]
+extern crate rental;
+#[cfg(feature = "fluent")]
+mod fluent;
+#[cfg(feature = "fluent")]
+pub use fluent::*;
+
+use carapax::{core::types::Update, Context, HandlerFuture, HandlerResult, UpdateHandler};
+use std::collections::HashMap;
+
+/// An i18n handler for carapax
+#[derive(Debug, Clone)]
+pub struct I18nHandler<R, T> {
+    resolver: R,
+    default: T,
+    languages: Vec<T>,
+}
+
+impl<R, T> I18nHandler<R, T> {
+    /// Creates new I18nHandler
+    ///
+    /// Search translator in languages or use default one
+    pub fn new<I>(resolver: R, default: T, languages: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+    {
+        Self {
+            resolver,
+            default,
+            languages: languages.into_iter().collect(),
+        }
+    }
+}
+
+impl<R, T> UpdateHandler for I18nHandler<R, T>
+where
+    R: LocaleResolver,
+    T: Translator + Send + Sync + 'static,
+{
+    fn handle(&self, context: &mut Context, update: &Update) -> HandlerFuture {
+        let lang = self.resolver.resolve(update);
+        let translator = self
+            .languages
+            .iter()
+            .find(|translator| {
+                lang.as_ref()
+                    .map(|lang| translator.locales().contains(lang))
+                    .unwrap_or(false)
+            })
+            .cloned()
+            .unwrap_or_else(|| self.default.clone());
+        context.set(translator);
+        HandlerResult::Continue.into()
+    }
+}
+
+/// An translator trait
+///
+/// Implement it to use your own i18n library
+pub trait Translator: Clone {
+    /// An error will be returned, if translation is failed
+    type Error;
+
+    /// Translates string with given arguments
+    fn translate(&self, args: &Arguments) -> Result<String, Self::Error>;
+
+    /// Returns available locales
+    fn locales(&self) -> Vec<String>;
+}
+
+/// A locale resolver trait
+pub trait LocaleResolver {
+    /// Resolves locale and returns [Translator](Translator)
+    fn resolve(&self, update: &Update) -> Option<String>;
+}
+
+/// Resolves translator by user's language code
+#[derive(Debug, Clone, Copy)]
+pub struct UserLocaleResolver;
+
+impl LocaleResolver for UserLocaleResolver {
+    fn resolve(&self, update: &Update) -> Option<String> {
```

Comment:
I think we can implement LocaleResolver for `Fn(update:&Update) -> Option<String>` and replace it with fn user_locale_resolver(...)

Reply:
I'd rather prefer to have a type rather than a free function... if I get what it's needed for )))